draft-ietf-dnsop-svcb-https.md

```diff
@@ -1082,18 +1082,37 @@ single-client granularity.
 
 ## Interaction with Alt-Svc
 
-Clients that do not implement support for Encrypted ClientHello MAY
-skip the HTTPS RR query
-if a usable Alt-Svc value is available in the local cache.
-If Alt-Svc connection fails, these clients SHOULD fall back to the HTTPS RR
-client connection procedure ({{client-behavior}}).
-
-Clients that implement support for ECH MUST perform the HTTPS RR query first,
-and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see
-{{ech-client-behavior}}).
-
-This specification does not alter the DNS records used when connecting
-to an Alt-Svc hostname (typically A and/or AAAA only).
+Clients that implement support for both Alt-Svc and HTTPS records SHOULD
+retrieve any HTTPS records for the Alt-Svc alt-authority, and ensure that
+their connection attempts are consistent with both the Alt-Svc parameters
+and any received HTTPS SvcParams.  If present, the HTTPS record's target
+and port override the alt-authority.  For example, suppose that
+"https://example.com" sends an Alt-Svc field value of:
+
+~~~ HTTP
+Alt-Svc: h2="alt.example.com:443", h3=":8443"
+~~~
+
+The client would retrieve the following HTTPS records:
+
+    alt.example.com. IN HTTPS 1 . alpn=h2,h3 ech=...
+    _8443._https.example.com. IN HTTPS 1 alt3.example.com. (
+        port=9443 alpn=h2,h3 ech=... )
```

Comment:
Probably good to have an incompatible example too. Something like...

Alt-Svc example:
```
Alt-Svc: h2="alt.example.com:443", h2="alt2.example.com:443", h3=":8443"
```

DNS example:
```
    alt.example.com. IN HTTPS 1 . alpn=h2,h3 ech=...
    alt2.example.com. IN HTTPS 1 . alpn=h3 ech=...
    _8443._https.example.com. IN HTTPS 1 alt3.example.com. (
        port=9443 alpn=h2,h3 ech=... )
```

New text:

> The client could then attempt an HTTP/3 connection to `alt3.example.com:9443` with ECH, or an HTTP over TLS connection to `alt.example.com:443` with ECH, as these options are consistent with both an Alt-Svc field value and its HTTPS record. It, however, could not attempt an HTTP over TLS connection to `alt2.example.com:443` as this option is not compatible with the Alt-Svc value's ALPN protocol.

Perhaps it's also worth something like:

> An incompatible SVCB option for an Alt-Svc value is considered to have failed. If all SVCB options for an Alt-Svc value fail, an SVCB-optional client ({{client-behavior}}) MAY fall back to a direct connection to the Alt-Svc value, unless prohibited by {{client-failures}} or {{ech-client-behavior}}. If all permitted SVCB or fallback options for an Alt-Svc value fail, the Alt-Svc connection is considered to have failed. The client MAY fall back to using the origin or another alternative service, as described in {{Section 2.4 of RFC7838}}.

Reply:
OK, I've added that example and incorporated various elements of this text.

Reply:
New text looks good to me.  Adds some nice clarity.

Some of the text on fallback to non-Alt-Svc gets a little bit close to impeding/redefining territory on RFC7838, but I think it's sufficiently on the safe side of the line, especially with the "as described in ..." reference.

Reply:
Hrm, I think this still doesn't quite capture it. The sentence ("An ECH-capable client would not connect [...]") seems to contain two steps at once. It's also misleading because non-ECH clients can also have see Alt-Svc vs ServiceForm incompatibilities. There's a difference between using ServiceMode records vs. A/AAAA if the alt2.example fetch gives back another TargetName. How about replacing:
```
    alt2.example.             IN HTTPS 1 . alpn=h3 ech=...
```
with:
```
    alt2.example.             IN HTTPS 1 alt2b.example. alpn=h3 ech=...
```

And then replace this sentence with:

> The client could not connect to `alt2b.example` because there is no possible connection that is consistent with both ALPN values ("h2" from Alt-Svc, "h3" from SvcParams). A SVCB-reliant client could not use the `alt2.example` Alt-Svc entry, but an SVCB-optional client could instead connect to `alt2.example` without the use of ServiceMode records. In an ECH-capable client, the "ech" SvcParam triggers SVCB-reliant client behavior ({{ech-client-behavior}}).

Reply:
Could break up the 2 things completely, and stop trying to contain the SVCB-reliant stuff just to the alt2 example.  One sentence about how the client could not use the HTTPS info from alt2.example (or state this via the alt2b.example stuff).  Then another sentence saying something along the lines of "For all 3 of these names, an ECH-capable client could not fall back to plain A/AAAA-based connection because the "ech" SvcParam triggers SVCB-reliant client behavior ({{...}})."

Reply:
Ah yeah, good point. The A/AAAA fallback case is relevant for both incompatible Alt-Svc/SVCB and plain connection error. I think that means we should also incorporate some more text from my original suggestion. The aim was not to be prescriptive about the order, but to translate these new notions of incompatibility and expanded Alt-Svc connection attempts into the existing notions of failure and fallback. I think we should spell out the general rule here.

`alt2b.example` isn't as important, but without it, there is no difference for an ECH-incapable client between the A/AAAA version and the ServiceForm version. Introducing a TargetName makes them different, which makes it clearer that this actually makes a difference.

Reply:
I added the `alt2b.example` TargetName and restructured this section as an enumeration instead of prose.  Please review.